### ISSUE TITLE
#[deprecated] lint doesn't trigger when overriding deprecated method

### ISSUE THREAD

**Issue by m-ou-se** [rust-lang/rust#98990]
This doesn't give any warnings, even though std::error::Error::description is deprecated:
```rust
impl std::error::Error for E {
    fn description(&self) -> &str {
        ":)"
    }
}
```

It seems like this is a deliberate decision:

https://github.com/rust-lang/rust/blob/3dcb616888aac50d55160b025266d555dad937d9/compiler/rustc_passes/src/stability.rs#L763-L764

Substituting that first `None` for `Some(impl_item_ref.id.hir_id())` makes it work.

Doing so results in the lint triggering in exactly one place in `std`:

https://github.com/rust-lang/rust/blob/3dcb616888aac50d55160b025266d555dad937d9/library/std/src/error.rs#L605-L609

That's because in all other cases, we have `#[allow(deprecated)]` there:

https://github.com/rust-lang/rust/blob/3dcb616888aac50d55160b025266d555dad937d9/library/std/src/error.rs#L597-L602

Which is suprising, as that lint doesn't tigger there, making that `#[allow]` unnecessary.

It seems as if it used to be necessary in the past, but now no longer is.

Did #[deprecated] trigger on overrides in the past? Was that behavior changed on purpose? Or is it a bug that it doesn't do that anymore?

**Comment by m-ou-se** [comment]
cc @rust-lang/libs-api, should the #[deprecated] lint trigger here or not?

**Comment by eddyb** [comment]
*(reposting https://github.com/rust-lang/rust/pull/98991#issuecomment-1177257829 here as well)*

I'm not sure what happened here, but the lack of tests for this is worrying (and may explain how this happened, if it was accidental).

**Comment by thomaseizinger** [comment]
I just ran into this. I'd like to guide my users away from implementing one trait function to new ones in a backwards-compatible way. My thinking was:

- Add new trait functions with default implementations that call the old one.
- Deprecate the old function.

I was expecting to have users (the trait provides a plugin-like functionality, so has many external users outside the crate) to see a deprecation warning that they should no longer implement this function but a new one instead.

**Comment by joshtriplett** [comment]
Sorry for the *long* delay in this being talked about!

We talked about this in today's @rust-lang/lang meeting, and those present (myself, @traviscross, and (representing libs-api) @Amanieu) were all generally in favor. We did feel this needs an FCP, though.

Please do provide a PR implementing this, mark it I-lang-nominated, and we'll FCP it.
